File: sdks/typescript/examples/end-to-end.ts
import {
  ControlFrame,
  REGISTRY_BUNDLE_MANIFEST_REQUEST_SUBJECT,
  registryBundleManifestRequestFrame,
  validateRegistryBundleManifestResponse,
} from "../src/index.ts";

const frame = registryBundleManifestRequestFrame({
  requirePublication: true,
  requireDrivers: true,
});
const encoded = frame.encode();
const parsed = ControlFrame.decode(encoded);

if (parsed.subject !== REGISTRY_BUNDLE_MANIFEST_REQUEST_SUBJECT) {
  throw new Error(`unexpected subject ${parsed.subject}`);
}

validateRegistryBundleManifestResponse(
  {
    schema_version: 1,
    node_id: "00000000-0000-0000-0000-000000000001",
    unavailable_reason: "example has no network peer",
  },
  parsed.jsonBody() as { schema_version: number; require_publication: boolean; require_drivers: boolean },
);

console.log(`built ${parsed.subject} (${encoded.byteLength} bytes)`);

File: sdks/rust/examples/end_to_end.rs
use zap_sdk::{
    ControlFrame, RegistryBundleManifestRequest, registry_bundle_manifest_request_frame,
};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let frame = registry_bundle_manifest_request_frame(true, true)?;
    let encoded = frame.encode();
    let parsed = ControlFrame::decode(&encoded)?;
    let body: RegistryBundleManifestRequest = parsed.json_body()?;

    println!(
        "built {} ({} bytes, require_drivers={})",
        parsed.subject(),
        encoded.len(),
        body.require_drivers
    );
    Ok(())
}
